clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -523,6 +626,21 @@ private static int parseAcks(String acksString) {
      * Implementation of asynchronously send a record to a topic.
      */
     private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback callback) {
+        if (transactionState != null && transactionState.isTransactional() && !transactionState.hasPid()) {
+            if (transactionState.isFenced()) {
+                throw Errors.INVALID_PRODUCER_EPOCH.exception();
```

Comment:
I'm not sure we want to throw the static exception since we won't get a stack trace (at least currently).